mlrun/runtimes/base.py

```diff
@@ -1121,10 +1170,84 @@ def _is_runtime_resource_run_in_transient_state(
 
         return False, last_update
 
+    def _verify_run_reached_stable_state_and_monitor_progress(
+        self, db: DBInterface, db_session: Session, project: str, run_uid: str
+    ):
+        """
+        This function runs in a retry_until_successful and therefore need to raise if run didn't reach stable state
+        """
+        desired_run_state = self._resolve_run_state_representing_current_resource_state(
+            db, db_session, project, run_uid
+        )
+        updated_run_state = self._ensure_run_status_updated(
+            db, db_session, project, run_uid, desired_run_state
+        )
+
+        if updated_run_state not in RunStates.stable_states():
+            raise Exception("run did not reach stable state yet")
+
+    def _resolve_run_state_representing_current_resource_state(
+        self, db: DBInterface, db_session: Session, project: str, run_uid: str
+    ) -> str:
+        crd_group, crd_version, crd_plural = self._get_crd_info()
+        if crd_group and crd_version and crd_plural:
+            crd_object = self._get_run_crd_object(project, run_uid)
+            (_, _, desired_run_state,) = self._resolve_crd_object_status_info(
+                db, db_session, crd_object
+            )
+        else:
+            pod = self._get_run_pod(project, run_uid)
+            (_, _, desired_run_state,) = self._resolve_pod_status_info(
+                db, db_session, pod
+            )
+
+        return desired_run_state
+
+    def _get_run_crd_object(self, project: str, run_uid: str):
+        label_selector = self._get_run_label_selector(project, run_uid)
+        k8s_helper = get_k8s_helper()
+        namespace = k8s_helper.resolve_namespace()
+        crd_group, crd_version, crd_plural = self._get_crd_info()
+        crd_objects = k8s_helper.crdapi.list_namespaced_custom_object(
+            crd_group,
+            crd_version,
+            namespace,
+            crd_plural,
+            label_selector=label_selector,
+        )
+        if not crd_objects or len(crd_objects["items"]) == 0:
+            raise RuntimeError("Run crd object could not be found")
+        if len(crd_objects["items"]) > 1:
+            logger.warning(
+                "Unexpectedly received more than one crd object for run. Best effort - using the first one"
```

Comment:
```suggestion
                "Received more than one crd record for run. Heuristically using the first one"
```
Can this happen? should this not be a real severe error?

Reply:
Not really, I mean, if this happens it means that two CRD objects somehow got the `mlrun/uid` label with the same value, which is very weird, question is what should monitoring do ? just give up and leave this run to whatever the state currently is ?

Reply:
This function is not used anymore (as part of moving to one loop to monitor all runs), but since I think it may be useful in the future I kept it